File: apps/web/lib/events.ts
/**
 * Typed events flowing over the LiveKit data channel between
 * the Python agent and the Next.js frontend.
 *
 * Direction is annotated as `// agent -> ui` or `// ui -> agent`.
 */

// ---------- agent -> ui ----------
export type StepChangeEvent = {
  type: "step.change";
  step: "greet" | "consent" | "pan" | "qa" | "confirm" | "verify" | "offer" | "ended";
};

export type ConsentRequestEvent = {
  type: "consent.request";
  consent_type: string;
  prompt: string;
};

export type PanRequestEvent = {
  type: "pan.request";
  prompt: string;
};

export type SignalsUpdateEvent = {
  type: "signals.update";
  signals: {
    stt?: string;
    vision?: string;
    geo?: string;
    fraud?: string;
    cibil?: string | number;
    risk?: string | number;
    [k: string]: string | number | undefined;
  };
};

export type CaptionEvent = {
  type: "caption";
  speaker: "drishti" | "customer";
  text: string;
  is_final: boolean;
  /** Server-stamped millis at emit time — UI sorts by this. */
  ts_ms?: number;
};

export type OfferTier = {
  tier: "conservative" | "standard" | "stretch";
  amount: number;
  rate_pct: number;
  tenure_months: number;
  emi: number;
  processing_fee: number;
  total_cost_of_credit: number;
};

export type DecisionType = "offer" | "soft_decline" | "hard_decline" | "human_review";

export type ConfirmedProfile = {
  age: number;
  monthly_income: number;
  employment_type: string;
  loan_purpose: string;
  requested_amount: number;
  declared_city: string;
};

export type ProfileConfirmRequestEvent = {
  type: "profile.confirm.request";
  profile: ConfirmedProfile;
  profile_version: number;
};

export type OfferShowEvent = {
  type: "offer.show";
  decision: DecisionType;
  offers?: OfferTier[];
  offer_version?: number;
  reason?: string;
  next_best_action?: string;
  shap_top3?: { feature: string; impact: number }[];
};

export type FraudFlagEvent = {
  type: "fraud.flag";
  signal: string;
  severity: number;
  reason: string;
};

export type SessionEndedEvent = {
  type: "session.ended";
  outcome: "approved" | "declined" | "fraud_block" | "human_review";
  audit_hash?: string;
  selected_offer?: (OfferTier & { offer_version?: number }) | null;
};

export type UiAckEvent = {
  type: "ui.ack";
  event_id: string;
  ok: boolean;
};

export type StateSnapshotEvent = {
  type: "state.snapshot";
  step: StepChangeEvent["step"];
  offer?: {
    decision?: DecisionType | null;
    offers?: OfferTier[];
    offer_version?: number;
    reason?: string;
    next_best_action?: string;
    shap_top3?: { feature: string; impact: number }[];
  };
  selected_offer?: (OfferTier & { offer_version?: number }) | null;
  ended?: { outcome?: SessionEndedEvent["outcome"]; audit_hash?: string } | null;
};

// ---------- ui -> agent ----------
export type ConsentGivenEvent = {
  type: "consent.given";
  event_id?: string;
  consent_type: string;
  spoken_text: string;
};

export type PanUploadedEvent = {
  type: "pan.uploaded";
  event_id?: string;
  pan_number: string;
  name: string;
  dob: string;
  photo_data_url: string;
  live_photo_data_url: string;
};

export type OfferSelectedEvent = {
  type: "offer.selected";
  event_id?: string;
  tier: "conservative" | "standard" | "stretch";
  offer_version?: number;
};

export type ProfileConfirmedEvent = {
  type: "profile.confirmed";
  event_id?: string;
  profile_version: number;
  accepted: boolean;
};

export type GeoReportEvent = {
  type: "geo.report";
  lat: number;
  lng: number;
  city?: string;
};

export type StateRequestEvent = {
  type: "state.request";
  event_id?: string;
};

// ---------- union ----------
export type AgentEvent =
  | StepChangeEvent
  | ConsentRequestEvent
  | PanRequestEvent
  | ProfileConfirmRequestEvent
  | SignalsUpdateEvent
  | CaptionEvent
  | OfferShowEvent
  | FraudFlagEvent
  | SessionEndedEvent
  | UiAckEvent
  | StateSnapshotEvent;

export type UiEvent =
  | ConsentGivenEvent
  | PanUploadedEvent
  | OfferSelectedEvent
  | ProfileConfirmedEvent
  | GeoReportEvent
  | StateRequestEvent;

// ---------- helpers ----------
export function encodeEvent(e: AgentEvent | UiEvent): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(e));
}

export function decodeEvent(payload: Uint8Array): AgentEvent | UiEvent | null {
  try {
    const text = new TextDecoder().decode(payload);
    return JSON.parse(text);
  } catch {
    return null;
  }
}

File: apps/agent/drishti_agent/events.py
"""
Typed events flowing over the LiveKit data channel.
Mirrors apps/web/lib/events.ts on the agent side.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field, asdict
from typing import Any, Literal


# ---------------- agent -> ui ----------------
@dataclass
class StepChange:
    type: Literal["step.change"] = "step.change"
    step: str = "greet"


@dataclass
class CaptionEvent:
    speaker: Literal["drishti", "customer"] = "drishti"
    text: str = ""
    is_final: bool = True
    # Server-side wall-clock millis at the moment the caption is emitted.
    # The UI sorts captions by this so even if reliable-data delivery shuffles
    # arrival order, the displayed sequence matches the spoken sequence.
    ts_ms: int = 0
    type: Literal["caption"] = "caption"


@dataclass
class SignalsUpdate:
    signals: dict[str, Any] = field(default_factory=dict)
    type: Literal["signals.update"] = "signals.update"


@dataclass
class PanRequest:
    prompt: str = "Please upload a clear photo of your PAN card."
    type: Literal["pan.request"] = "pan.request"


@dataclass
class ConsentRequest:
    consent_type: str = "data_processing"
    prompt: str = ""
    type: Literal["consent.request"] = "consent.request"


@dataclass
class ProfileConfirmRequest:
    profile: dict[str, Any] = field(default_factory=dict)
    profile_version: int = 0
    type: Literal["profile.confirm.request"] = "profile.confirm.request"


@dataclass
class OfferShow:
    decision: str = "offer"
    offers: list[dict] = field(default_factory=list)
    offer_version: int = 0
    reason: str | None = None
    next_best_action: str | None = None
    shap_top3: list[dict] = field(default_factory=list)
    type: Literal["offer.show"] = "offer.show"


@dataclass
class FraudFlag:
    signal: str = ""
    severity: int = 2
    reason: str = ""
    type: Literal["fraud.flag"] = "fraud.flag"


@dataclass
class SessionEnded:
    outcome: str = "approved"
    audit_hash: str | None = None
    selected_offer: dict[str, Any] | None = None
    type: Literal["session.ended"] = "session.ended"


@dataclass
class UiAck:
    event_id: str = ""
    ok: bool = True
    type: Literal["ui.ack"] = "ui.ack"


@dataclass
class StateSnapshot:
    step: str = "greet"
    offer: dict[str, Any] = field(default_factory=dict)
    selected_offer: dict[str, Any] | None = None
    ended: dict[str, Any] | None = None
    type: Literal["state.snapshot"] = "state.snapshot"


# ---------------- helpers ----------------
def encode(event) -> bytes:
    """Serialize an event dataclass for publishing on the data channel.

    `ensure_ascii=False` so the rupee symbol and Hindi captions survive the
    round trip without being mangled into \\uXXXX escapes.
    """
    return json.dumps(
        asdict(event), separators=(",", ":"), ensure_ascii=False
    ).encode("utf-8")


def decode(payload: bytes) -> dict | None:
    try:
        return json.loads(payload.decode("utf-8"))
    except Exception:
        return None
